File: go/day1/day1.go
package day1

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

type Direction uint64

const (
	Left Direction = iota
	Right
)

func (d Direction) String() string {
	switch d {
	case Left:
		return "Left"
	case Right:
		return "Right"
	default:
		return "Unknown"
	}
}

func parseDirection(s string) Direction {
	switch s {
	case "L":
		return Left
	case "R":
		return Right
	default:
		panic("unreachable: invalid direction")
	}
}

type Command struct {
	value     uint64
	direction Direction
}

type Lock struct {
	value uint64
}

func (d *Lock) left() {
	if d.value == 0 {
		d.value = 99
	} else {
		d.value -= 1
	}
}

func (d *Lock) right() {
	d.value = (d.value + 1) % 100
}

func Run(part2 bool) {
	dat, err := os.ReadFile("./day1/input.txt")
	if err != nil {
		panic(err)
	}

	src := string(dat)
	src = strings.TrimSpace(src)

	lines := strings.Split(src, "\n")

	commands := make([]Command, len(lines))
	for i, line := range lines {
		direction := parseDirection(line[:1])
		value_str := line[1:]

		value, err := strconv.ParseUint(value_str, 10, 64)
		if err != nil {
			panic(err)
		}

		commands[i] = Command{value, direction}
	}

	lock := Lock{
		value: 50,
	}

	count := 0

	if part2 {
		for _, command := range commands {
			for range command.value {
				switch command.direction {
				case Left:
					lock.left()
				case Right:
					lock.right()
				}

				if lock.value == 0 {
					count += 1
				}
			}
		}
	} else {
		for _, command := range commands {
			for range command.value {
				switch command.direction {
				case Left:
					lock.left()
				case Right:
					lock.right()
				}
			}

			if lock.value == 0 {
				count += 1
			}
		}
	}

	fmt.Println(count)
}

File: zig/src/day1/day1.zig
const std = @import("std");

const Direction = enum {
    left,
    right,
};

const Command = struct {
    value: u32,
    direction: Direction,

    fn parse(buf: []const u8) !Command {
        const direction = switch (buf[0]) {
            'L' => Direction.left,
            'R' => Direction.right,
            else => @panic("Unknown direction"),
        };
        const value = try std.fmt.parseInt(u32, buf[1..], 10);

        return Command{
            .direction = direction,
            .value = value,
        };
    }
};

const Lock = struct {
    value: u32 = 50,

    fn left(self: *Lock) void {
        if (self.value == 0) {
            self.value = 99;
        } else {
            self.value -= 1;
        }
    }

    fn right(self: *Lock) void {
        self.value = (self.value + 1) % 100;
    }
};

pub fn run(part2: bool) !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    defer _ = gpa.detectLeaks();

    const alloc = gpa.allocator();

    var lock = Lock{};

    const file_data = try std.fs.cwd().readFileAlloc(alloc, "./src/day1/input.txt", std.math.maxInt(usize));
    defer alloc.free(file_data);

    var count: u32 = 0;

    var it = std.mem.splitScalar(u8, file_data, '\n');
    while (it.next()) |x| {
        const trimmed = std.mem.trim(u8, x, " \n");
        if (trimmed.len == 0) continue;

        const command = try Command.parse(trimmed);

        if (part2) {
            for (0..command.value) |_| {
                switch (command.direction) {
                    Direction.left => lock.left(),
                    Direction.right => lock.right(),
                }

                if (lock.value == 0) {
                    count += 1;
                }
            }
        } else {
            for (0..command.value) |_| {
                switch (command.direction) {
                    Direction.left => lock.left(),
                    Direction.right => lock.right(),
                }
            }

            if (lock.value == 0) {
                count += 1;
            }
        }
    }

    std.debug.print("count: {d}\n", .{count});
}
